### PR TITLE
[Wsign-compare] warning resolutions, by directory, 5

### PR DESCRIPTION
**Pull request by tg-at-google**
Resolution affecting directories:
`tensorflow/compiler/xla/`

### CODE REVIEW

File: tensorflow/compiler/xla/client/lib/pooling.cc [hunk 1]
```diff
@@ -198,15 +198,16 @@ XlaOp AvgPoolGrad(XlaOp out_backprop, absl::Span<const int64> gradients_size,
   XlaBuilder* b = out_backprop.builder();
   return b->ReportErrorOrReturn([&]() -> StatusOr<XlaOp> {
     const int num_dims = kernel_size.size();
-
-    if (gradients_size.size() != num_dims) {
+    const int gradients_size_size = gradients_size.size();
+    if (gradients_size_size != num_dims) {
```
Comment (mihaimaruseac) [inline tensorflow/compiler/xla/client/lib/pooling.cc:202]: Please use a readable name. How about `num_gradients`?

File: tensorflow/compiler/xla/client/lib/pooling.cc [hunk 2]
```diff
@@ -198,15 +198,16 @@ XlaOp AvgPoolGrad(XlaOp out_backprop, absl::Span<const int64> gradients_size,
...
       return tensorflow::errors::InvalidArgument("gradients must be ", num_dims,
                                                  "-dimensional");
     }
 
     TF_ASSIGN_OR_RETURN(Shape out_backprop_xla_shape,
                         b->GetShape(out_backprop));
-    if (out_backprop_xla_shape.dimensions().size() != num_dims) {
+    const int obxsd_size = out_backprop_xla_shape.dimensions().size();
+    if (obxsd_size != num_dims) {
```
Comment (mihaimaruseac) [inline tensorflow/compiler/xla/client/lib/pooling.cc:210]: Same as above, name should be readable. How about `num_dimensions`?

File: tensorflow/compiler/xla/client/xla_builder.cc [hunk 3]
```diff
@@ -766,15 +766,17 @@ XlaOp XlaBuilder::BroadcastInDim(
...
+    tensorflow::int64 broadcast_dimensions_size = broadcast_dimensions.size();
+    if (operand_shape->rank() != broadcast_dimensions_size) {
       return InvalidArgument(
           "Size of broadcast_dimensions has to match operand's rank; operand "
           "rank: %lld, size of broadcast_dimensions %u.",
           operand_shape->rank(), broadcast_dimensions.size());
     }
-    for (int i = 0; i < broadcast_dimensions.size(); i++) {
+    for (int i = 0, end = broadcast_dimensions.size(); i < end; i++) {
+      const tensorflow::int64 out_dim_size_size = out_dim_size.size();
```
Comment (mihaimaruseac) [inline tensorflow/compiler/xla/client/xla_builder.cc:777]: Please use readable name.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/client/lib/pooling.cc (hunk 1 above)
Comment (tg-at-google) [reply tensorflow/compiler/xla/client/lib/pooling.cc:202]: updated.

File: tensorflow/compiler/xla/client/lib/pooling.cc (hunk 2 above)
Comment (tg-at-google) [reply tensorflow/compiler/xla/client/lib/pooling.cc:210]: updated.

File: tensorflow/compiler/xla/client/xla_builder.cc (hunk 3 above)
Comment (tg-at-google) [reply tensorflow/compiler/xla/client/xla_builder.cc:777]: updated.

**Comment by tg-at-google** [comment]
updates applied.

File: tensorflow/compiler/xla/client/xla_builder.cc [hunk 4]
```diff
@@ -3021,7 +3024,12 @@ StatusOr<XlaOp> XlaBuilder::AddInstruction(HloInstructionProto&& instr,
     instr.add_operand_ids(operand.handle());
   }
 
-  *instr.mutable_metadata() = metadata_;
+  if (one_shot_metadata_.has_value()) {
+    *instr.mutable_metadata() = one_shot_metadata_.value();
+    one_shot_metadata_.reset();
+  } else {
+    *instr.mutable_metadata() = metadata_;
+  }
```
Comment (mihaimaruseac) [inline tensorflow/compiler/xla/client/xla_builder.cc:3032]: Unrelated change

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by tg-at-google** [comment]
updates applied, 2.

**Comment by tg-at-google** [comment]
updates applied, 2.

File: tensorflow/compiler/xla/client/xla_builder.cc [hunk 5]
```diff
@@ -766,15 +766,17 @@ XlaOp XlaBuilder::BroadcastInDim(
     TF_ASSIGN_OR_RETURN(auto output_shape,
                         ShapeUtil::MakeValidatedShape(
                             operand_shape->element_type(), out_dim_size));
-    if (operand_shape->rank() != broadcast_dimensions.size()) {
+    tensorflow::int64 broadcast_dimensions_size = broadcast_dimensions.size();
+    if (operand_shape->rank() != broadcast_dimensions_size) {
```
Comment (mihaimaruseac) [inline tensorflow/compiler/xla/client/xla_builder.cc:770]:
```suggestion
    tensorflow::int64 broadcast_rank = broadcast_dimensions.size();
    if (operand_shape->rank() != broadcast_rank) {
```

File: tensorflow/compiler/xla/client/xla_builder.cc [hunk 6]
```diff
@@ -766,15 +766,17 @@ XlaOp XlaBuilder::BroadcastInDim(
...
       return InvalidArgument(
           "Size of broadcast_dimensions has to match operand's rank; operand "
           "rank: %lld, size of broadcast_dimensions %u.",
           operand_shape->rank(), broadcast_dimensions.size());
     }
-    for (int i = 0; i < broadcast_dimensions.size(); i++) {
+    for (int i = 0, end = broadcast_dimensions.size(); i < end; i++) {
```
Comment (mihaimaruseac) [inline tensorflow/compiler/xla/client/xla_builder.cc:776]:
```suggestion
    for (int i = 0; i < broadcast_rank; i++) {
```

(we already have the local)

File: tensorflow/compiler/xla/client/xla_builder.cc [hunk 7]
```diff
@@ -786,7 +788,7 @@ XlaOp XlaBuilder::BroadcastInDim(
                            *operand_shape, output_shape, broadcast_dimensions)
                            .status());
     std::vector<int64> in_dim_size(out_dim_size.begin(), out_dim_size.end());
-    for (int i = 0; i < broadcast_dimensions.size(); i++) {
+    for (int i = 0, end = broadcast_dimensions.size(); i < end; i++) {
```
Comment (mihaimaruseac) [inline tensorflow/compiler/xla/client/xla_builder.cc:791]:
```suggestion
    for (int i = 0; i < broadcast_rank; i++) {
```

(same here)

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by tg-at-google** [comment]
updates applied, 3.

**Review by mihaimaruseac — APPROVED** [review]

File: tensorflow/compiler/xla/literal.cc [hunk 8]
```diff
@@ -1891,13 +1894,13 @@ bool LiteralBase::IsR1Iota() const {
...
+        return Get<uint8>({idx}) == static_cast<uint8>(idx);
       case U16:
-        return Get<uint16>({idx}) == idx;
+        return Get<uint16>({idx}) == static_cast<uint16>(idx);
       case U32:
-        return Get<uint32>({idx}) == idx;
+        return Get<uint32>({idx}) == static_cast<uint32>(idx);
       case U64:
-        return Get<uint64>({idx}) == idx;
+        return Get<uint64>({idx}) == static_cast<uint64>(idx);
```
Comment (mihaimaruseac) [inline tensorflow/compiler/xla/literal.cc:1903]: https://godbolt.org/z/sxThad says that we should instead `static_cast<int64>(Get<...>)` instead of the RHS.

Didn't notice the bug until now, this might be why builds fail

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by tg-at-google** [comment]
update applied, 4.

File: tensorflow/compiler/xla/literal.cc [hunk 9]
```diff
@@ -1891,13 +1894,13 @@ bool LiteralBase::IsR1Iota() const {
   auto is_iota_at_idx = [&](const int64 idx) {
     switch (shape().element_type()) {
       case U8:
-        return Get<uint8>({idx}) == idx;
+        return static_cast<uint8>(Get<uint8>({idx})) == idx;
```
Comment (mihaimaruseac) [inline tensorflow/compiler/xla/literal.cc:1897]: Should be cast to `int64`, as that is the type of `idx`

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/literal.cc (hunk 9 above)
Comment (tg-at-google) [reply tensorflow/compiler/xla/literal.cc]: updated.

**Comment by tg-at-google** [comment]
update applied, 5.

**Review by mihaimaruseac — APPROVED** [review]
